dart/dynamics/Entity.h

```diff
@@ -63,11 +63,23 @@ class Shape;
 /// may have different policies about how/if their reference frame or name
 /// can be changed. Use the Detachable class to create an Entity whose reference
 /// Frame can be changed arbitrarily.
-class Entity
+class Entity : public common::Subscription
 {
 public:
   friend class Frame;
 
+  enum Notification {
```

Comment:
Your event concept does look like a good replacement for the general-purpose notification sending. The only thing I'd be concerned about is destruction notifications, which was the primary motivation behind the subscription concept: Inheritance can guarantee that the destruction notification gets sent out when the object is destructed, and it can guarantee that this notification will occur for any class that inherits the Publisher class.

The destruction notification *could* be done through an Event object, but it would require (1) Creating an Event object for it, and (2) Making sure the Event is triggered during destruction (and *only* during destruction). Also, for the templated sub_ptr class to work, there would be a third requirement (3) The destruction Event object is accessed in the same exact way for every class that wants to work with sub_ptr. On the other hand, using inheritance for the specific purpose of destruction notifications would remove all these requirements, and simply involve inheriting the Publisher class.

So I think I would propose a compromise: continue to use inheritance for destruction notifications, but get rid of the current general-purpose notification system in favor of using the Event class. I think this compromise would simultaneously maximize safety, efficiency, and generalizability. If we do this, we might want to change the names of the Subscriber and Publisher classes to be more specific to the fact that they're only tracking destruction.